### ISSUE TITLE
Traceroute service type broken 

### ISSUE THREAD

**Issue by marian-babik** [perfsonar/meshconfig-admin#38]
Hi,
MCA relies on the "service type" concept to organise the test meshes and currently there are service types defined such as pinger, traceroute, latency, bandwidth. This requires a certain mapping to take place to understand which nodes can participate in which type of tests, which I guess has been usually done by extracting this information from SLS. Below is a screenshot of hostgroup config showing service type field.
<img width="1297" alt="screen shot 2018-02-16 at 2 41 22 pm" src="https://user-images.githubusercontent.com/9883414/36310351-011614fa-1328-11e8-832f-75a666a935e1.png">

I'm not sure if this is configuration issue of our MCA instance or a potential bug, but it looks like that recently this mapping has broken as most of the hosts that are synced from global SLS lookup no longer declare to be able to run traceroute, thus effectively we don't have a way to configure traceroute testing for them (as they don't show up in the list of hosts for traceroute service type). It could be also that something has changed in SLS.

Below are screenshots of two examples, perfsonar01-iep-grid.saske.sk host which comes from our custom local SLS - showing traceroute under services

<img width="831" alt="screen shot 2018-02-16 at 2 40 28 pm" src="https://user-images.githubusercontent.com/9883414/36310375-14be0742-1328-11e8-974c-4a61c6b44b8b.png">

and atlas*.bu.edu which uses global SLS and doesn't show traceroute under supported services.

<img width="907" alt="screen shot 2018-02-16 at 2 40 02 pm" src="https://user-images.githubusercontent.com/9883414/36310380-181b235c-1328-11e8-8570-44d3370471e9.png">

**Comment by soichih** [comment]
@marian-babik 

It's been a while since I troubleshoot-ed MCA, so my memory is a bit rusty.. but I will do my best.

So I can confirm that, there is no traceroute service listed for perfsonar-cms2.itns.purdue.edu https://meshconfig.grid.iu.edu/#!/hosts/58f74b23f5139f0021ac2a04 which causes meshconfig publisher to not list this host in the generated /us-cms meshconfig. I've roughly looked and it looks like none of the hosts that information comes from SLS are listed in meshconfig, is that what you are seeing also? 

For `perfsonar-cms2.itns.purdue.edu`, the host page shows that the information comes from [this SLS instance](http://monipe-ls.rnp.br:8090/lookup/host/a5a33c0a-fe48-40cd-a2c5-8b64f45a57ca) So, I queried all services listed under [this host](http://monipe-ls.rnp.br:8090/lookup/records?service-host=lookup/host/a5a33c0a-fe48-40cd-a2c5-8b64f45a57ca) and I see following services

```
    "service-type": [      "bwctl"    ],
    "service-type": [      "mp-bwctl"    ],
    "service-type": [      "mp-owamp"    ],
    "service-type": [      "ma"    ],
    "service-type": [      "pscheduler"    ],
```

which isn't the exact match as listed under MCA's service endpoint, so I might be missing something, but I wonder if SLS used to show service-type: ["traceroute"] in the past, but not anymore? 

@apertome Do you know of any changes (within the last few month) on how registration service registers traceroute service? I am not even sure if there ever was such a thing as "traceroute service(server)".. so maybe we can just add "traceroute" service on all hosts?

**Comment by ShawnMcKee** [comment]
I think putting a "traceroute" service on all hosts could work.  We want to be able to initiate traceroutes on any hosts running bwctl or pscheduler so maybe add it to any host with either of those services?   This bug is actually high-priority in that we are no longer getting traceroute measurements for hosts coming in via SLS.

**Comment by apertome** [comment]
@ShawnMcKee The changes to the traceroute registration in the LS happened around a year ago, so I don't understand why this would just now manifest itself. Are you sure nothing has changed in your instance?

This is definitely something I intend to fix before the 4.1 beta release, but I don't see doing a new release between now and then -- the entire deployment model is changing, so it wouldn't make sense to figure out how to do a new release to publish this one fix.

However, I will review the code and see if I can see an obvious/easy fix. If so, I may be able to provide a patch.

**Comment by apertome** [comment]
Additional context from Andy:
> I don’t remember when we last registered a separate traceroute service. I
> think it must pre-date 3.5.1 because I can’t find an example of the
> lsregistrationdaemon.conf file that contains it. We should not be
> registering a “traceroute” service because there is no such thing. If a
> host record exists, you should be allowed to setup a traceroute test in the
> GUI. Same for ping. I know at one time we did have records for those, but
> they were dropped because they did not give useful information.
> 
> Generally speaking I don’t really like the GUI limiting the user to being
> able to setup certain types of tests because its not in the LS. The LS
> should be used as a guideline to help you set things up, but making it a
> strict requirement kills usability IMO. This is especially true in this
> case because a traceroute service is non-sensical, but I believe holds true
> for other types of tests where an admin may know more than the LS.
> 
> You might want to double-check the GUI isn't looking for BWCTL in order to
> setup throughput tests because that is not accurate either since a
> pScheduler host without BWCTL works just as well.

**Comment by ShawnMcKee** [comment]
@apertome I think it could have been happening for a while and we didn't notice.  Most hosts are inserted from the OIM/GOCDB and they don't show the problem.   Many hosts took a while to upgrade (we *still* have some 3.x hosts!).     It would be good if there was a simple patch you add to allow us to include those hosts in traceroute tests.   Otherwise we are missing any topology context for the measurements we are making with those hosts.

**Comment by marian-babik** [comment]
We just found out that global SLS record likely has priority over OIM/GOCDB record in MCA, so all hosts registered in global SLS will be missing traceroute tests (traceroute test won't be added into auto-mesh json since the node doesn't support traceroute service). So the impact is larger then we initially thought, it could impact all/most of the OSG hosts and probably some GOCDB hosts as well.

**Comment by apertome** [comment]
I still have not had a chance to look at this, and can't get to it this week. Furthermore, I'll be out of the country the next two weeks for the perfSONAR developer meeting and other travels, so I would not expect any movement on this in the next couple weeks.

The way the MCA was designed, it considers the LS as the ultimate source of truth, which doesn't always play out well in practice (like in this situation).

We are planning on reducing reliance on the LS in general in the future; this is a great example of why we want to do that. 

I'm still hoping to be able to come up with a simple patch for this issue, but not sure when I will be able to look at it.

**Comment by soichih** [comment]
@apertome I believe it should be very simple to have traceroute service added to all LS ingested sites. Since you are busy, I will take a look and send a PR.

**Comment by apertome** [comment]
That's very kind, @soichih , I'm not saying I think that will be complicated, but part of the issue is that I have not entirely figured out/tested the build/release/upgrade process for this yet. As well as figuring out how to merge just this change, without the many others that have been made. If you create a PR I can look at that when I have time, and would also hopefully work as a patch for Shawn/Marian in the short term ... if you want to do that, feel free. If not, I will look at it when I have the time. Either way, thanks for your help!

**Comment by ShawnMcKee** [comment]
Thanks @soichih !
Marian and I can put any PR you produce in place (temporarily) in meshconfig.grid.iu.edu just to fix the current issue.   Longer term we can work with @apertome to get the official version deployed when it is ready.   Thanks for your help!
Shawn

**Comment by soichih** [comment]
@apertome I've submitted a PR. Please take a look.

@ShawnMcKee 
If Michael is going to create a new docker container, then you can just docker pull it on meshconfig.grid.iu.edu, or you can manually apply the patch - since it really involves [a single line change](https://github.com/perfsonar/meshconfig-admin/pull/40/files#diff-1bb4a8139b2fa6521445d05022390a0aR66)

> We are planning on reducing reliance on the LS in general in the future;

Does this mean LS would be deprecated eventually? MCA was designed with assumption that information on LS would improve as time goes by. How can users reliably find hosts that they can initiate tests to? Maybe a list from [the directory?](http://perfsonar-jax-level3-01.loni.org/toolkit/gui/perfAdmin/directory.cgi)

By the way, on the directory page, I see "Traceroute Daemons". What are they? I also found "traceroute" service listed in LS 

http://ps-west.es.net:8090/lookup/records?service-host=lookup/host/a8080e52-aa25-4542-914a-bea206d5c244
http://ps-west.es.net:8090/lookup/records?service-host=lookup/host/79e65307-83e5-474a-8115-dd6312312443
http://ps-east.es.net:8090/lookup/records?service-host=lookup/host/d8e0d463-d7e6-466b-a780-8285802cbde6

**Comment by apertome** [comment]
Thanks Soichi,

No, I won't be doing an immediate release; I'd recommend if Shawn wants to patch his instance that he do it manually.

Re: traceroute daemons, I don't know. Maybe Andy knows? @arlake228

**Comment by apertome** [comment]
No, the LS is not going away, but there are quite a few deployments that don't use the LS at all, and right now, it's difficult to schedule tests to hosts that are NOT in the LS. 

This specific issue is actually a good example of the kinds of LS dependencies that we want to avoid.

I reviewed your PR, but it is more complicated than you indicated, and conflicts with some changes I've made; I'll need to test and merge it carefully.
